### PR TITLE
feat: Add support for feedback events on chatbot responses #2165

### LINKED ISSUE #2165: Like/Dislike and Events for ChatBot
#### Is your feature request related to a problem? Please describe
I would like my users to be able to share feedback about responses from chat bots. 

#### Describe the solution you'd like

Feedback: I would like to optionally turn on feedback for the GenAI responses in the form of a Thumbs Up / Thumbs Down buttons for each chat response that could be new events. As a developer, I would want to know the text and the button clicked (thumbs up or down). I would record the feedback but not make any change to the UI other than I would want the button to look selected.

#### Describe alternatives you've considered

(A clear and concise description of any alternative solutions or features you've considered.)

### PR DESCRIPTION
**Pull request by marek-mihok**
**The PR fulfills these requirements:** (check all the apply)

- [x] It's submitted to the `main` branch.
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `feat: Add a button #xxx`, where "xxx" is the issue number).
- [x] When resolving a specific issue, the PR description includes `Closes #xxx`, where "xxx" is the issue number.
- [x] If changes were made to `ui` folder, unit tests (`make test`) still pass.
- [x] New/updated tests are included

___

Usage example:

https://github.com/h2oai/wave/assets/23740173/1ff4870e-73f5-4d82-b053-384a0cb4ab0f

This PR also features an API change - added a `feedback` event:
```tsx
/** Create a chatbot card to allow getting prompts from users and providing them with LLM generated answers. */
export interface Chatbot {
  /** An identifying name for this component. */
  name: Id
  /** Chat messages data. Requires list or cyclic buffer. */
  data: Rec
  /** Chat input box placeholder. Use for prompt examples. */
  placeholder?: S
  /** The events to capture on this chatbot. One of 'stop' | 'scroll_up' | 'feedback'. */
  events?: S[]
  /** True to show a button to stop the text generation. Defaults to False. */
  generating?: B
  /** The previous messages to show as the user scrolls up. */
  prev_items?: ChatbotMessage[]
}
```

Closes #2165

### CODE REVIEW

File: ui/src/chatbot.tsx [hunk 1]
```diff
@@ -91,12 +97,14 @@ export interface Chatbot {
...
   placeholder?: S
-  /** The events to capture on this chatbot. One of 'stop' | 'scroll_up'. */
+  /** The events to capture on this chatbot. One of 'stop' | 'scroll_up' | 'feedback'. */
   events?: S[]
   /** True to show a button to stop the text generation. Defaults to False. */
   generating?: B
   /** The previous messages to show as the user scrolls up. */
   prev_items?: ChatbotMessage[]
+  /** True to show thumbs up/down buttons to capture the feedback event on chatbot response. Defaults to False. */
+  feedback?: B
```
Comment (mturoci) [inline ui/src/chatbot.tsx:107]: Isn't registering `feedback` event enough?

File: website/widgets/ai/chatbot.md [hunk 2]
```diff
@@ -63,3 +63,22 @@ q.page['example'] = ui.chatbot_card(
     events=['scroll']
 )
 ```
+
+## With a feedback
```
Comment (mturoci) [inline website/widgets/ai/chatbot.md:67]:
```suggestion
## Collect feedback
```

File: ui/src/chatbot.tsx [hunk 3]
```diff
@@ -180,8 +204,20 @@ export const XChatbot = (props: Chatbot) => {
               paddingBottom: msgs?.[idx + 1]?.from_user !== from_user ? rem(0.8) : 0,
               color: from_user ? '$text' : botTextColor
             }} >
-            <span className={clas(css.msg, 'wave-s14')} style={{ padding: content?.includes('\n') ? 12 : 6 }}>
+            <span className={clas(css.msg, 'wave-s14')} style={{ padding: content?.includes('\n') ? 12 : 12, paddingBottom: props.feedback && !from_user ? 30 : 0 }}>
               <Markdown source={content || ''} />
+              {props.feedback && !from_user &&
+                <div className={css.feedback}>
+                  <Fluent.IconButton
+                    iconProps={{ iconName: (rated.has(idx) && rated.get(idx)) ? 'LikeSolid' : 'Like' }}
```
Comment (mturoci) [inline ui/src/chatbot.tsx:212]: What is the `get` for?

```suggestion
                    iconProps={{ iconName: rated.has(idx) ? 'LikeSolid' : 'Like' }}
```

File: ui/src/chatbot.tsx [hunk 4]
```diff
@@ -180,8 +204,20 @@ export const XChatbot = (props: Chatbot) => {
               paddingBottom: msgs?.[idx + 1]?.from_user !== from_user ? rem(0.8) : 0,
               color: from_user ? '$text' : botTextColor
             }} >
-            <span className={clas(css.msg, 'wave-s14')} style={{ padding: content?.includes('\n') ? 12 : 6 }}>
+            <span className={clas(css.msg, 'wave-s14')} style={{ padding: content?.includes('\n') ? 12 : 12, paddingBottom: props.feedback && !from_user ? 30 : 0 }}>
```
Comment (mturoci) [inline ui/src/chatbot.tsx:207]: This line is odd: `content?.includes('\n') ? 12 : 12`. Also, does it look good for both single line and multiline cases?

File: ui/src/chatbot.tsx [hunk 5]
```diff
@@ -106,6 +114,7 @@ export const XChatbot = (props: Chatbot) => {
     [msgs, setMsgs] = React.useState<Message[]>(props.data ? processData(props.data) : []),
     [userInput, setUserInput] = React.useState(''),
     [isInfiniteLoading, setIsInfiniteLoading] = React.useState(false),
+    [rated, setRated] = React.useState(() => new Map()),
```
Comment (mturoci) [inline ui/src/chatbot.tsx:117]: Why is this a function? The map could also use some types. Do we need an external map though? Couldn't the  feedback state be part of the message instead?

File: ui/src/chatbot.tsx [hunk 6]
```diff
@@ -134,7 +143,22 @@ export const XChatbot = (props: Chatbot) => {
         setIsInfiniteLoading(true)
       }
     }, [props.events, props.name]),
-    onDataChange = React.useCallback(() => { if (props.data) setMsgs(processData(props.data)) }, [props.data])
+    onDataChange = React.useCallback(() => { if (props.data) setMsgs(processData(props.data)) }, [props.data]),
+    handleFeedback = (positive: B, id: I) => {
+      if (props.events?.includes('feedback')) {
```
Comment (mturoci) [inline ui/src/chatbot.tsx:148]: Isn't this already taken care of within the render logic? The buttons shouldn't even render if this is not true, right?

File: ui/src/chatbot.tsx [hunk 7]
```diff
@@ -134,7 +143,22 @@ export const XChatbot = (props: Chatbot) => {
         setIsInfiniteLoading(true)
       }
     }, [props.events, props.name]),
-    onDataChange = React.useCallback(() => { if (props.data) setMsgs(processData(props.data)) }, [props.data])
+    onDataChange = React.useCallback(() => { if (props.data) setMsgs(processData(props.data)) }, [props.data]),
+    handleFeedback = (positive: B, id: I) => {
```
Comment (mturoci) [inline ui/src/chatbot.tsx:147]: Let's go for 2 separate handleNegative, handlePositive instead.

File: ui/src/chatbot.tsx [hunk 8]
```diff
@@ -134,7 +143,22 @@ export const XChatbot = (props: Chatbot) => {
...
+      if (props.events?.includes('feedback')) {
+        setRated(rated => {
+          if (rated.has(id) && rated.get(id) === positive) {
```
Comment (mturoci) [inline ui/src/chatbot.tsx:150]: Wouldn't this be enough?
```suggestion
          if (rated.get(id) === positive) {
```

File: ui/src/chatbot.tsx [hunk 9]
```diff
@@ -70,6 +71,11 @@ const
       marginBottom: 7.5,
       transform: 'translateX(-50%)',
       width: 180
+    },
+    feedback: {
```
Comment (mturoci) [inline ui/src/chatbot.tsx:75]: Why does this need to be absolutely positioned?

**Review by mturoci — CHANGES_REQUESTED** [review]
Thanks @marek-mihok!

File: ui/src/chatbot.tsx (hunk 1 above)
Comment (marek-mihok) [reply ui/src/chatbot.tsx:107]: Makes sense, it would be easier in terms of usage.

File: ui/src/chatbot.tsx (hunk 9 above)
Comment (marek-mihok) [reply ui/src/chatbot.tsx:75]: A leftover from right placed buttons implementation.

File: ui/src/chatbot.test.tsx [hunk 10]
```diff
@@ -159,4 +159,72 @@ describe('XChatbot', () => {
...
+    expect(dislikeButton).toBeInTheDocument()
+  })
+
+  it('Fires a like feedback event when clicked on the thumbs up button', () => {
+    const { container } = render(<XChatbot {...{ ...model, data, events: ['feedback'] }} />)
+    const likeButton = container.querySelector("i[data-icon-name='Like']") as HTMLLIElement
+
+    fireEvent.click(likeButton)
+    expect(emitMock).toHaveBeenCalled()
+    expect(emitMock).toHaveBeenCalledTimes(1)
```
Comment (mturoci) [inline ui/src/chatbot.test.tsx:177]: Nit, no need for explicit check.

```suggestion
    expect(emitMock).toHaveBeenCalledTimes(1)
```

File: ui/src/chatbot.test.tsx [hunk 11]
```diff
@@ -159,4 +159,72 @@ describe('XChatbot', () => {
...
+    fireEvent.click(dislikeButton)
+    expect(emitMock).toHaveBeenCalled()
+    expect(emitMock).toHaveBeenCalledTimes(1)
+    expect(emitMock).toHaveBeenCalledWith(model.name, 'feedback', { message: data[1].content, positive: false })
+
+    const dislikeSolidButton = container.querySelector("i[data-icon-name='DislikeSolid']") as HTMLLIElement
+    expect(dislikeSolidButton).toBeInTheDocument()
+  })
+
+  it('Fires a no feedback event when clicked on the thumbs up button twice', () => {
```
Comment (mturoci) [inline ui/src/chatbot.test.tsx:197]: Like/Dislike should be fired at most once unless like => dislike => like etc.

File: ui/src/chatbot.tsx [hunk 12]
```diff
@@ -134,7 +139,23 @@ export const XChatbot = (props: Chatbot) => {
         setIsInfiniteLoading(true)
       }
     }, [props.events, props.name]),
-    onDataChange = React.useCallback(() => { if (props.data) setMsgs(processData(props.data)) }, [props.data])
+    onDataChange = React.useCallback(() => { if (props.data) setMsgs(processData(props.data)) }, [props.data]),
+    handlePositive = (id: I) => {
+      setMsgs(messages => {
+        const isPositive = messages[id]?.positive
+        messages[id].positive = isPositive ? undefined : true
```
Comment (mturoci) [inline ui/src/chatbot.tsx:146]: Should never be undefined. Same for the other handler.

**Review by mturoci — CHANGES_REQUESTED** [review]

File: ui/src/chatbot.test.tsx (hunk 11 above)
Comment (marek-mihok) [reply ui/src/chatbot.test.tsx:197]: So undoing of like/dislike shouldn't be possible?

Comment (mturoci) [reply ui/src/chatbot.test.tsx:197]: Correct.

**Review by mturoci — APPROVED** [review]
LGTM. Thanks @marek-mihok!

**Pull request merged**